apps/firelens-stability/lib/cloud/ecs.ts

```diff
@@ -41,86 +46,229 @@ async function processEcsTestTaskQueue() {
+    }
+
+    // Launch tasks in groups of 10 or less
+    console.log(`  🧪 ${testCase.managed.collectionName}/${testCase.managed.suiteName}/${testCase.managed.caseName}: launched tasks`);
+    const taskCount = testCase.config.taskCount;
+    let launchedTasks = await launchECSTasks(testCase, taskCount, ecs, taskDefinitionArn);
+
+    // Add to the back of the task queue after a delay of 5-10 seconds for validation. Check on the status.
+    addECSTaskToQueue(
+        10000, 20000, {
+            successPromiseResolver: task.successPromiseResolver,
+            executionRecord: {
+                startTime: Date.now(),
+                taskArns: launchedTasks,
+            },
+            testCase: testCase,
+            executionRound: task.executionRound + 1,
+            taskDefinitionArn: taskDefinitionArn,
+        });
+}
+
+function addECSTaskToQueue(delayBase: number, delayCap: number, ecsTask: IEcsTestTask) {
+    setTimeout(() => {
+        ecsTestCaseTaskQueue.push(ecsTask);
+    }, delayBase + Math.random() * (delayCap - delayBase));
+}
+
+function ecsTaskReturn(ecsTestTask: IEcsTestTask) {
+    ecsTestTask.successPromiseResolver(
+        /* Prune away the ecsTestTask specific information */
+        {
+            executionRecord: ecsTestTask.executionRecord,
+            testCase: ecsTestTask.testCase,
         }
+    );
+}
+
+async function validateAndRetryECSTask(ecsTestTask: IEcsTestTask) {
+    const testCase = ecsTestTask.testCase;
+      
+    // Set the region name
+    AWS.config.update({ region: ecsTestTask.testCase.config.region });
     
-        // Get the task definition file
-        const taskDefinition = await getTestCaseTaskDefinition(testCase);
+    // Create an ECS client
+    const ecs = new AWS.ECS();
 
+    const {
+        currentTasks,
+        startingTasks,
+        runningTasks,
+        startingTasksCount,
+        runningTasksCount,
+        endedTasksCount,
+    } = await validateECSTestTask(ecs, ecsTestTask);
+
+    /* Done! All tasks are in running */
+    if (runningTasksCount === ecsTestTask.testCase.config.taskCount) {
+        console.log(`  ✅ ${testCase.managed.collectionName}/${testCase.managed.suiteName}/${testCase.managed.caseName}: All tasks in running state`);
+
+        /* Mark the ECS task as resolved. */
+        ecsTaskReturn(ecsTestTask);
+        return;
+    }
+    
+    /* Tasks need to be retried. Retry the ended tasks */
+    let taskArns = ecsTestTask.executionRecord.taskArns;
+    if (endedTasksCount !== 0) {
+        console.log(`  🔁 ${testCase.managed.collectionName}/${testCase.managed.suiteName}/${testCase.managed.caseName}: Retrying failed tasks`);
+        const taskCount = endedTasksCount;
+        let newTasksLaunched = await launchECSTasks(testCase, taskCount, ecs, ecsTestTask.taskDefinitionArn);
+        taskArns = [...startingTasks, ...runningTasks, ...newTasksLaunched];
+    }
+
+    /* Add back to queue for validation */
+    addECSTaskToQueue(10000, 20000, {
```

Comment:
huge magic numbers: can these be constants and why is hte base so big?

Reply:
are these milliseconds?

Reply:
Updated.